### PR TITLE
Generate Windows That Conform To GeoTiff Segments

### PR DESCRIPTION
**Pull request by jamesmcclain**
- [x] Ensure that S3 window size is being properly interpreted (is `windowsSize` a length or an area in this context?)
- [x] Ensure that handling of striped segments is viable (are tiles of one pixel in height okay?)

Connects https://github.com/locationtech/geotrellis/issues/2173

### CODE REVIEW

File: s3/src/main/scala/geotrellis/spark/io/s3/S3GeoTiffRDD.scala [hunk 1]
```diff
@@ -132,28 +132,30 @@ object S3GeoTiffRDD extends LazyLogging {
     lazy val sourceGeoTiffInfo = S3GeoTiffInfoReader(bucket, prefix, options)
 
     (options.maxTileSize, options.partitionBytes) match {
-      case (None, Some(partitionBytes)) =>
-        val segments: RDD[(String, Array[GridBounds])] =
-          sourceGeoTiffInfo.segmentsByPartitionBytes(partitionBytes, windowSize)
+      case (_, Some(partitionBytes)) => {
+        val maxSize = math.min(options.maxTileSize.getOrElse(1<<10), windowSize.getOrElse(1<<10)) // XXX is windowSize a length or an area?
```
Comment (echeipesh) [inline s3/src/main/scala/geotrellis/spark/io/s3/S3GeoTiffRDD.scala]: `windowSize` is length, implicitly for square windows.

**Comment by echeipesh** [comment]
> Ensure that handling of striped segments is viable (are tiles of one pixel in height okay?)

It seems strange but the thinking went is that if you don't have an opinion about the shape of tiles you want you're probably using this RDD as input to tiler. From the perspective of optimizing for IO reading the segments as skinny tiles is fine, from the perspective of tiler chunking a skinny tile into multiple tiles is not problematic either.

**Comment by jamesmcclain** [comment]
> It seems strange but the thinking went is that if you don't have an opinion about the shape of tiles you want you're probably using this RDD as input to tiler. From the perspective of optimizing for IO reading the segments as skinny tiles is fine, from the perspective of tiler chunking a skinny tile into multiple tiles is not problematic either.

Sounds capital.

File: spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopGeoTiffRDD.scala [hunk 2]
```diff
@@ -87,22 +92,54 @@ object HadoopGeoTiffRDD {
...
-    val conf = configuration(path, options)
-    options.maxTileSize match {
-      case Some(tileSize) =>
+  def apply[I, K, V](
+    path: Path,
+    uriToKey: (URI, I) => K,
+    options: Options
+  )(implicit sc: SparkContext, rr: RasterReader[Options, (I, V)]): RDD[(K, V)] = {
+
+    val conf = new SerializableConfiguration(configuration(path, options))
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopGeoTiffRDD.scala]: Why `conf` should be wrapped here? Looks like it's only used for input formats arguments => can be used a common configuration without extra wrapper.

File: spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopGeoTiffRDD.scala [hunk 3]
```diff
@@ -87,22 +92,54 @@ object HadoopGeoTiffRDD {
...
+    val path2 = path.toString
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopGeoTiffRDD.scala]: Well it's always a bit confusing when you see in code such names. In fact you can write just 

```scala
HadoopGeoTiffInfoReader(path.toString, conf, options.tiffExtensions)
```

File: spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopGeoTiffRDD.scala [hunk 4]
```diff
@@ -129,18 +166,25 @@ object HadoopGeoTiffRDD {
...
-
     val windows: RDD[(Path, GridBounds)] =
       pathsToDimensions
         .flatMap { case (objectRequest, (cols, rows)) =>
-          RasterReader.listWindows(cols, rows, options.maxTileSize).map((objectRequest, _))
+          val info = HadoopGeoTiffInfoReader(objectRequest.toString, conf, options.tiffExtensions)
+          val layout = info.getGeoTiffInfo(objectRequest.toString).segmentLayout.tileLayout
+
+          RasterReader
+            .listWindows(cols, rows, options.maxTileSize.getOrElse(1<<10), layout.tileCols, layout.tileRows)
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopGeoTiffRDD.scala]: `maxTileSize` comment here too

File: s3/src/main/scala/geotrellis/spark/io/s3/S3GeoTiffRDD.scala [hunk 5]
```diff
@@ -205,7 +207,13 @@ object S3GeoTiffRDD extends LazyLogging {
     val windows =
       objectRequestsToDimensions
         .flatMap { case (objectRequest, (cols, rows)) =>
-          RasterReader.listWindows(cols, rows, options.maxTileSize).map((objectRequest, _))
+          val bucket = objectRequest.getBucketName
+          val key = objectRequest.getKey
+          val layout = sourceGeoTiffInfo.getGeoTiffInfo(s"s3://$bucket/$key").segmentLayout.tileLayout
+
+          RasterReader
+            .listWindows(cols, rows, options.maxTileSize.getOrElse(1<<10), layout.tileCols, layout.tileRows)
```
[the remark this answers is not in the document]
Comment (jamesmcclain) [reply s3/src/main/scala/geotrellis/spark/io/s3/S3GeoTiffRDD.scala]: As long as changing the default `maxTileSize` is not construed to constitute an API change, I prefer that.

File: spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopGeoTiffRDD.scala (hunk 2 above)
Comment (jamesmcclain) [reply spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopGeoTiffRDD.scala]: That variable makes its way into a `sc.parallelize` inside of `HadoopGeoTiffInfoReader`.

Comment (pomadchin) [reply spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopGeoTiffRDD.scala]: Good point, my bad.

**Comment by jamesmcclain** [comment]
All comments have been addressed, I believe.

**Comment by jamesmcclain** [comment]
:100:

**Pull request merged**